fms/utils/serialization.py

```diff
@@ -88,48 +107,101 @@ def get_adapted(
+        ) as model_weights:  # type: ignore[attr-defined]
+            return model_weights.get_tensor(key)
+
+
+class LazySafetensorsDict(collections.UserDict):
+    def set_lazy_tensor(self, key, file, device):
+        super().__setitem__(key, lambda: _get_safetensors_item(key, file, device))
+
+    def __getitem__(self, key):
+        lazy_tensor = super().__getitem__(key)
+        if callable(lazy_tensor):
+            lazy_tensor = lazy_tensor()
+            super().__setitem__(key, lazy_tensor)
+        return lazy_tensor
+
+
 def load_state_dict(
     model_path: Union[str, Path],
+    *,
+    source: Optional[str] = None,
     distributed_strategy: Optional[str] = None,
     checkpoint_sharding: Optional[str] = None,
     initial_device: torch.device = torch.device("cpu"),
     rank: int = 0,
     world_size: int = 1,
-) -> Mapping[str, Any]:
+) -> MutableMapping[str, Any]:
     """
     Validates that the file(s) found at a checkpoint path are compatible with
-    the intended (possibly distributed) use-case, and returns the state dict
-    if needed.
+    the intended (possibly distributed) use-case, and returns a lazy loading
+    state dict if possible (some formats may not support that).
+
+    If model_path is a directory, it'll try to load models based on the source
+    (e.g. .bin for HF, .pth for Meta), and, if no source is specified or hasn't
+    been registered, it'll try .safetensors, .pth, and .bin.
 
     Args:
     model_path: the path to find the weights. If not set, return None.
+    source: If the weights in the state dict didn't come from an FMS model,
+            `source` specifies which conversion function might be needed.
+            See `serialization.list_sources(architecture)`
     distributed_strategy: the kind of possibly-distributed model in which we
             intend to load these weights. E.g. tp, fsdp, None. Used for
             validation.
     checkpoint_sharding: the sharding format of the checkpoint.
             E.g. layer, tp, fsdp.
-    initial_device: where the state dict will be loaded. if meta, return None.
+    initial_device: where the state dict will be loaded if not lazy.
+            If meta, return empty dict.
     """
     if model_path is None or initial_device.type == "meta":
         return {}
-    # TODO: Add support for tp-sharding a non-sharded state dict.
-    if (
-        distributed_strategy == "tp" or checkpoint_sharding == "tp"
-    ) and distributed_strategy != checkpoint_sharding:
-        raise ValueError(
-            f"TP-sharded models are currently only compatible with TP-sharded checkpoints. Attempting to load {checkpoint_sharding} to {distributed_strategy}"
-        )
     if checkpoint_sharding == "fsdp" and distributed_strategy not in ["fsdp", "hsdp"]:
         raise ValueError(f"FSDP checkpoints can only be loaded into an FSDP model")
+    if checkpoint_sharding == "tp" and distributed_strategy not in ["tp"]:
```

Comment:
why "not in" instead of just testing for equality? there's only one option so seems like == makes more sense

Reply:
changing to equality